### PR TITLE
Restricts naming for repositories

### LINKED ISSUE #40817: Repository name with commas can't be retrieved
Add a repository with commas in name:

```
POST /_snapshot/test,with,comma
{
  "type" : "fs",
  "settings" : {
    "location" : "/tmp/es-backups"
  }
}
```

Attempting to retrieve information about this repository returns an error:
```
 GET /_snapshot/test,with,comma
```
or encoded
```
 GET /_snapshot/test%2Cwith%2Ccomma
```
returns
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "repository_missing_exception",
        "reason" : "[test] missing"
      }
    ],
    "type" : "repository_missing_exception",
    "reason" : "[test] missing"
  },
  "status" : 404
}
```

Could we disallow commas in repository name to prevent this from happening?

### PR DESCRIPTION
**Pull request by danhermann**
Applies the same naming restrictions to repositories as to snapshots except that leading underscores and uppercase characters are permitted.

Fixes #40817.

### CODE REVIEW

**Comment by danhermann** [comment]
Jenkins, test this please

**Comment by danhermann** [comment]
Jenkins, test this please

**Comment by danhermann** [comment]
Jenkins, test this please

**Review by original-brownbear — CHANGES_REQUESTED** [review]
@danhermann thanks! I think this is fine, can you add a quick test that checks these conditions as well please?

**Comment by original-brownbear** [comment]
@ywelsch maybe you can take a quick look too, to confirm that it's fine to restrict naming like this (looks like the same conditions as for snapshot naming are in here as well which makes perfect sense to me)?

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 1]
```diff
@@ -418,6 +420,26 @@ private Repository createRepository(RepositoryMetaData repositoryMetaData, Map<S
         }
     }
 
+    private static void validate(final String repositoryName) {
+        if (Strings.hasLength(repositoryName) == false) {
+            throw new RepositoryException(repositoryName, "cannot be empty");
+        }
+        if (repositoryName.contains(" ")) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: this is already covered by filename check

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 2]
```diff
@@ -418,6 +420,26 @@ private Repository createRepository(RepositoryMetaData repositoryMetaData, Map<S
...
+            throw new RepositoryException(repositoryName, "must not contain whitespace");
+        }
+        if (repositoryName.contains(",")) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: this is already covered by filename check

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 3]
```diff
@@ -402,6 +403,7 @@ private void closeRepository(Repository repository) {
      */
     private Repository createRepository(RepositoryMetaData repositoryMetaData, Map<String, Repository.Factory> factories) {
         logger.debug("creating repository [{}][{}]", repositoryMetaData.type(), repositoryMetaData.name());
+        validate(repositoryMetaData.name());
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: this is a method that is internally called both on repository creation, but also after a restart when a node rejoins a cluster with an existing repository. I think we should do the validation only for new repository creation requests, and not retroactively apply to old repositories (which will break them). This means that the validation should be done in `registerRepository`, not `createRepository`.

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left some comments. Thank you for contributing this PR.

**Comment by danhermann** [comment]
Jenkins, test this please

**Review by ywelsch — APPROVED** [review]
LGTM

**Review by original-brownbear — APPROVED** [review]
LGTM2

**Pull request merged**